Do not delete the service object until all tasks are successfully removed

nishanttotla opened the pull request:
This PR implements the proposal in https://github.com/docker/swarmkit/issues/2497#issuecomment-363602993.

A note on the task reaper logic, which should be carefully reviewed and might be slightly complex, but hopefully this description should make it clear:

- The task reaper is waiting on task update events, and when it sees a task with desired state `REMOVE` that has proceeded beyond `SHUTDOWN`, it tracks this task for removal.
- For each such task, the task reaper adds its service to `cleanupServices` if the service was also marked for removal.
- On each tick, services in `cleanupServices` are deleted if all their tasks are gone.

Close https://github.com/docker/swarmkit/issues/2497.

Signed-off-by: Nishant Totla <nishanttotla@gmail.com>

anshulpundir commented on manager/orchestrator/taskreaper/task_reaper.go:
```diff
@@ -282,6 +289,32 @@ func (tr *TaskReaper) tick() {
+		tr.store.View(func(tx store.ReadTx) {
+			for svc, _ := range tr.cleanupServices {
+				tasks, _ := store.FindTasks(tx, store.ByServiceID(svc))
+				if tasks != nil && len(tasks) == 0 {
```
What happens if tasks == nil ?

anshulpundir commented on manager/orchestrator/taskreaper/task_reaper.go:
```diff
@@ -282,6 +289,32 @@ func (tr *TaskReaper) tick() {
 	}
+
+	// Perform service cleanup
+	deleteServices := make(map[string]struct{})
```
Should the trigger for service cleanup instead be task status update event ?

nishanttotla replied:
I think it's okay to do cleanups on ticks, rather than every task update. For a service that has a large number of tasks, maybe it's too much unnecessary actions we'd be taking.

anshulpundir replied:
Makes sense

anshulpundir commented on manager/orchestrator/taskreaper/task_reaper.go:
```diff
@@ -282,6 +289,32 @@ func (tr *TaskReaper) tick() {
+		tr.store.View(func(tx store.ReadTx) {
+			for svc := range tr.cleanupServices {
+				tasks, _ := store.FindTasks(tx, store.ByServiceID(svc))
+				if tasks != nil && len(tasks) == 0 {
```
What if tasks == nil ?

nishanttotla replied:
It seems like this doesn't happen from the scenarios I tested (when there are no tasks and the service is still around. I also looked at the code for `FindTasks` and it seems reasonable), but I'll double check.

anshulpundir replied:
Yep, I guess tasks shouldn't be nil. Does it make sense to either remove the nil check, or add a condition where tasks ==  nil is also an indication to clean up ?

nishanttotla replied:
Actually it seems like `len(tasks)` will be 0 even if `tasks` is `nil`. The right thing to do is probably to not discard the error.

nishanttotla commented on api/objects.proto:
```diff
@@ -122,6 +122,13 @@ message Service {
+	// API, and the orchestrator is waiting for all the tasks to
+	// be properly shut down before removing the service object
+	// from the store.
+	bool removed = 7;
```
Noted.

nishanttotla commented:
Removing the WIP on this PR. I believe that the allocator test failure is flakiness/unrelated because I get these on the master branch too when I test locally.

Ping @docker/core-swarmkit-maintainers

anshulpundir commented on manager/orchestrator/taskreaper/task_reaper.go:
```diff
@@ -282,6 +289,39 @@ func (tr *TaskReaper) tick() {
+				}
+				if len(tasks) == 0 {
+					deleteServices[svc] = struct{}{}
+					delete(tr.cleanupServices, svc)
```
Instead of removing one by one, just clear out the map at the end.

nishanttotla replied:
I do this because I want to keep `cleanupServices` around. It won't necessarily be empty after this loop is run.

cyli replied:
Should this delete happen later, after the batch service delete?  i.e. After the batch is successful, delete all the services in `deleteServices` from `tr.cleanupServices`.

If the batch fails, then `tr.cleanupServices` will no longer have a reference to the service, and the service may not be updated again so would it be possible that the service never gets cleaned up?

anshulpundir commented on api/objects.proto:
```diff
@@ -122,6 +122,13 @@ message Service {
+	// API, and the orchestrator is waiting for all the tasks to
+	// be properly shut down before removing the service object
+	// from the store.
+	bool marked_removed = 7;
```
I hate to nit pick on this, but marked_removed is a little confusing. Maybe marked_for_removal or pending_removal ?

nishanttotla replied:
I don't mind updating it. Maybe let's wait to see if others have any comments, then I can update it once at the end.

nishanttotla commented:
@anshulpundir on second thoughts I don't know if listening to service update events is desirable. Most of the cleanup code for a service will only run if it was scaled down or removed, as opposed to having to do store lookups each time the service was updated.

anshulpundir commented:
How do you plan to add the services for cleanup in the taskreaper when the service has been scaled down to zero tasks ?

dperny commented:
Yeah, good catch, seems to me that the case for service deletion of a 0-task service isn't handled.

nishanttotla commented:
@dperny @anshulpundir I'm aware of that case, sorry I should have pointed out in the PR description. So my thought was that in such a case, a service can be deleted right away (in controlapi). Waiting for the task reaper to do it is not required, and also will need for us to search though all services, which we would like to avoid.

Question is, do you think it is acceptable?

anshulpundir commented:
I think it is somewhat inelegant/hanky to delete services in two different components.

nishanttotla commented:
I agree it's not elegant. I also don't want to listen on service events in the task reaper, so ideas are welcome.

anshulpundir commented:
True, that’s pretty inelegant too.

anshulpundir commented:
Ok, I vote going ahead with deleting directly in the control api for now. We can change it if a better idea comes forward.

nishanttotla commented:
Changed name of field to `MarkedForRemoval`, and now removing replicated services with zero instances in the control api right away. Also rebased to have one commit.

nishanttotla commented on manager/controlapi/service.go:
```diff
@@ -839,8 +839,28 @@ func (s *Server) RemoveService(ctx context.Context, request *api.RemoveServiceRe
+			if int64(m.(*api.ServiceSpec_Replicated).Replicated.Replicas) == 0 {
+				return store.DeleteService(tx, request.ServiceID)
+			}
+		case *api.ServiceSpec_Global:
```
@anshulpundir @dperny suggestions for whether or not we want to handle this case here. Not sure what the right condition would be here, but for checking nodes in the cluster or something like that.

anshulpundir replied:
Can we just get tasks by service ?

nishanttotla replied:
Sorry for the long silence on this PR. What happens if say the cluster is a small number of nodes, and all tasks restart around the same time for some reason? This is probably a corner case that may not be critical. The issue here is that the service spec for global services isn't enough to make this decision unlike replicated services.

anshulpundir replied:
Sorry, I'm not sure I follow. Lets talk in person.

anshulpundir commented on manager/controlapi/service.go:
```diff
@@ -843,8 +843,28 @@ func (s *Server) RemoveService(ctx context.Context, request *api.RemoveServiceRe
+		m := service.Spec.GetMode()
+		switch m.(type) {
+		case *api.ServiceSpec_Replicated:
+			if int64(m.(*api.ServiceSpec_Replicated).Replicated.Replicas) == 0 {
```
Replicated.Replicas is the replication target. I think there can be a case when the target is 0 but the actual tasks have not exited ?

cyli commented on manager/orchestrator/global/global_test.go:
```diff
@@ -470,7 +479,8 @@ func updateService(t *testing.T, s *store.MemoryStore, service *api.Service, for
 func deleteService(t *testing.T, s *store.MemoryStore, service *api.Service) {
 	s.Update(func(tx store.Tx) error {
-		assert.NoError(t, store.DeleteService(tx, service.ID))
+		service.MarkedForRemoval = true
```
Would it make sense to do a `store.GetService`, then update the service, as is done for `updateService`, and not setting the `MarkedForRemoval` flag to false everywhere?  

That way you don't accidentally set the `MarkedForRemoval` flag on the original service object.  I'm mainly a little wary of blanket-setting the `MarkedForRemoval` flag to false in all these tests, because that may cover up a potential bug where the field should not have been set.

cyli commented on manager/orchestrator/taskreaper/task_reaper.go:
```diff
@@ -37,18 +37,37 @@ type TaskReaper struct {
+// addServiceForCleanup checks if the service is marked for removal, and if it is,
+// adds it to the map cleanupServices so that it can be considered for deletion later.
+// This is only done if the service doesn't already exist in cleanupServices.
+func (tr *TaskReaper) addServiceForCleanup(id string) {
```
Is this function still used?

nishanttotla replied:
Nope, will remove it. I must have forgotten.

cyli commented on manager/orchestrator/taskreaper/task_reaper_test.go:
```diff
@@ -161,5 +221,9 @@ func TestTaskReaperInit(t *testing.T) {
 		assert.NotNil(t, store.GetTask(tx, "terminaltask1"))
 		// and the second terminal task should have been removed
 		assert.Nil(t, store.GetTask(tx, "terminaltask2"))
+		// the service that was marked for removal should be gone
```
Should `serviceMarkedForRemovalNoTasks` also be added to this list?

nishanttotla commented:
@cyli 

> Should this delete happen later, after the batch service delete? i.e. After the batch is successful, delete all the services in `deleteServices` from `tr.cleanupServices`.

Good point. It seems like we don't do this for deleting tasks either. This is not an issue in the case when the leader failed and the node went down, because on starting the task reaper, we make a fresh list of services marked for removal anyway. If the batch failed for another reason though, it's possible that we'll be in a bad state. I can fix this in a follow up perhaps, because it'll need to be changed for deletion of tasks as well.

nishanttotla commented on manager/allocator/allocator_test.go:
```diff
@@ -409,6 +411,8 @@ func TestAllocator(t *testing.T) {
 		s := store.GetService(tx, "testServiceID1")
 		s.Spec.Endpoint = nil
 
+		// ensuer that the service isn't removed by another test
```
I think these can be removed. In the previous case, we were just deleting the service, so it not expected that we'll be using removed services afterwards.

nishanttotla commented:
@cyli I've addressed your comments.

I agree with you that reusing service objects in unit tests is a risk. For now, I have removed all but one of the instances of setting `MarkedForRemoval` to `false` (this is the case for creating a new service in the store, so we're fine there).

anshulpundir commented on manager/allocator/allocator.go:
```diff
@@ -156,7 +156,6 @@ func (a *Allocator) init(ctx context.Context, aa *allocActor) (<-chan events.Eve
 		api.EventDeleteNetwork{},
 		api.EventCreateService{},
 		api.EventUpdateService{},
-		api.EventDeleteService{},
```
How was the allocator previously using this event ?

anshulpundir replied:
I see that the change to not act on EventDeleteService in allocator/network.go is in a different commit. I'd suggest keeping it in the same commit. If you're planning to merge them, please ignore my comment.

anshulpundir commented on manager/orchestrator/taskreaper/task_reaper.go:
```diff
@@ -282,6 +307,39 @@ func (tr *TaskReaper) tick() {
+				// with length zero. If an error occurs though, we log it and don't mark the service
+				// for removal yet.
+				if err != nil {
+					log.G(context.Background()).WithError(err).Errorf("error in task reaper while looking up tasks for service %s", svc)
```
lets pass in the context from the task reaper instead of using context.Background()

anshulpundir commented on manager/orchestrator/taskreaper/task_reaper.go:
```diff
@@ -282,6 +307,39 @@ func (tr *TaskReaper) tick() {
+				tasks, err := store.FindTasks(tx, store.ByServiceID(svc))
+				// if all tasks of svc have been cleaned up already, then we should get a response
+				// with length zero. If an error occurs though, we log it and don't mark the service
+				// for removal yet.
```
So how does this service get removed ? please clarify the comment.

anshulpundir commented on manager/orchestrator/taskreaper/task_reaper.go:
```diff
@@ -282,6 +307,39 @@ func (tr *TaskReaper) tick() {
+					log.G(context.Background()).WithError(err).Errorf("error in task reaper while looking up tasks for service %s", svc)
+					continue
+				}
+				if len(tasks) == 0 {
```
len(tasks) may not be zero. Consider the case of deleting a service whose tasks were already shutdown. Another case is when you restart the reaper and there are services from previous term which need to be deleted but may have services to be removed. You'll need to check the desired and actual state of each task if the len(tasks) > 0

nishanttotla replied:
I agree, this will need to be fixed.

nishanttotla replied:
@anshulpundir I've added a fix for this. This required another change that tasks of the service that was being removed needed to be newly tracked for removal before the tick was finished. PTAL.

nishanttotla commented:
@anshulpundir pointed out that another case where we may have an issue is
- `service rm` command passed
- leadership changes before desired state of all tasks is set to `REMOVE`

When the new leader starts, there are no events pertaining to any of the tasks of this service (or the service itself), which means the service will never shut down. The fix for this is that when the orchestrator starts, it needs to re-initiate shutdown of any services that are in this state.

cyli commented:
Nice catch @anshulpundir!

nishanttotla commented:
Currently due for this change:

- [ ] Make sure that the orchestrator handles services with `MarkedForRemoval` set to `true` on startup (in this PR)
- [ ]  Rename task reaper (follow-up)
- [ ] Don't clear cleanup lists/maps until the deletion batch is successful for tasks/services (follow-up)
